0000-error-values-proposal.rst

```diff
@@ -0,0 +1,260 @@
+-------------------
+
+With the approach described in this proposal, the different error types
+would be scattered accross the different components of the GHC library,
+there would therefore not be a single place to look at to see all the
+errors. They would have to look for all the places in the code where we
+pack some error in a ``GHCError`` wrapper.
+
+Another drawback is that error consumers have to resort to using
+``Typeable`` to implement specific behaviours for any given type, instead
+of pattern matching on constructors of a large sum type of all possible
+errors. Both of those drawbacks come from the dynamic component of the
+design put forward in this proposal, which is essential to avoid the
+import cycles mentionned earlier. We consider the import cycles to be
+a more serious drawback than the dynamic errors, especially given that
+the number of concrete error/warning types should be quite reasonable.
+
+The major cost of implementing this proposal is the sheer amount of
+refactoring that will be necessary to emit error values and move the
+rendering to much later.
+
+Alternatives
+------------
+
+As mentionned in the previous section, we considered a similar approach
+but where there is no existential wrapper and the errors from different
+subsystems are just glued together using sum types. That would have
+required importing all subsystems in one place, introducing import cycles
+that would be very delicate to handle and incur too expensive a price on
+all GHC developers. Another problem (small, in comparison to the
+previous one) is that those intermediate "glue" sum types introduce
+some maintenance overhead, because of the need to enumerate all types of
+errors that GHC can throw, at various points of the module/subsystem
+hierarchy, as we make the error types bubble up until we define a
+toplevel error ADT that can represent any error that GHC can throw.
+
+On the other end of the spectrum, we could also define one type per
+error and just pack them all into an existential wrapper with suitable
+dictionaries. That would have the advantage of not requiring any
+"glue" but would be very painful for consumers, as they would not
+have any convenient way to handle a whole group of errors that are
+somewhat related, e.g coming from the same subsystem. This would be
+particularly problematic if we end up wanting to introduce a level or
+two in the hierarchy between the existential wrapper at the top and a
+concrete sum of errors for a given part of GHC (a leaf).
+
+Unresolved Questions
+--------------------
+
+We have not fully fleshed out the forest of error types and believe this
+is something that will be best done by scanning GHC's code, looking for
+functions that emit error messages and trying to adapt them to emit an
+existentially-wrapped, subsystem-specific error **value**.
+
+Implementation Plan
+-------------------
+
+Well-Typed LLP will implement this proposal with financial support from
+Richard Eisenberg.
```

Comment:
IIRC this should mention the specific grant from which this funding came.

Reply:
I just copied the previous proposal's footer. I'm not sure I ever saw the specific grant "identifier" for this project, but I'd be happy to add it as soon as you or @goldfirere provide it to me.

Reply:
NSF grant number 1704041.

Reply:
Done, thanks!